### DISCUSSION TITLE
자바 파일이 실행되는 과정을 설명해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#39] [Java]
.

**Comment by kunsanglee** [comment]
1. 소스코드 작성 
- .java 확장자를 가진 파일에 소스 코드를 작성합니다.
2. 컴파일
- 작성된 소스 코드를 자바 컴파일러(javac)에 의해 컴파일 됩니다. 컴파일러는 소스 코드를 분석하고, 문법적 오류를 검사한 후, 바이트코드(.class 파일)로 변환합니다.
3. 로딩
- JVM은 필요한 .class 파일들을 로드합니다. 이 과정에서 클래스 로더(Class Loader)가 사용되며, 로딩된 클래스들은 JVM의 메로리 영역인 메소드 영역(Method Area)에 배치됩니다.
4. 링킹 
- 로딩된 바이트코드는 링킹 과정을 거칩니다. 링킹은 검증(Verify), 준비(Prepare), 해석(Resolve)의 세 단계로 이루어집니다. 이 과정에서 바이트코드의 유효성 검사, 메모리 할당 및 상수 풀 참조 등이 이루어집니다.
5. 초기화 
- 링킹이 완료된 후, 클래스 초기화 단계에서는 정적 변수의 초기화 및 정적 초기화 블록(static initializer blocks)이 실행됩니다. 이 과정은 클래스가 처음 사용될 때 한 번만 수행됩니다.
6. 실행 
- 초기화가 완료되면 JVM은 main 메서드를 호출하여 프로그램을 실행합니다. 이 때 자바 인터프리터는 .class 파일에 있는 바이트 코드를 한 줄씩 읽어 실행합니다. 필요에 따라 JVM 내장의 JIT(Just-In-Time) 컴파일러가 바이트코드를 더 빠르게 실행할 수 있는 기계어 코드로 변환하기도 합니다.
7. 실행 종료
- 프로그램이 종료되면, 사용된 자원들은 운영체제에 의해 회수되고, JVM은 종료됩니다.

**Comment by seunghye218** [comment]
먼저 작성한 .java 파일이 컴파일러를 통해 전달된 소스 코드를 바이트 코드라고 하는 기계 독립적인 코드로 인코딩합니다. 각 내용은 별도의 .class 파일에 저장됩니다.
    
클래스 로더를 통해 프로그램에서 참조되는 다른 모든 클래스가  로드됩니다.
    
바이트 코드 검증자가 Java 인터프리터에 전달된 코드가 실행하기에 적합한 상태이고 Java 인터프리터가 손상될 염려 없이 실행할 수 있는지 확인합니다. 
검증 작업으로 변수는 사용되기 전에 초기화야하고, 접근 제어 규칙을 위반하지 않는지 등이 있습니다.  
    
그 다음 바이트코드 인터프리터와 JIT 컴파일러가 바이트코드를 기계어 코드로 변환하여 하드웨어가 코드를 실행하게 됩니다.

설명한 프로세스로 인해 Java 프로그램은 대상 운영 체제에 독립적이지만 실행 시간이 훨씬 더 깁니다.


![image](https://github.com/2024-woowacourse-study/level-interview/assets/93831492/72212385-8cac-478a-a852-aea6c1f04ce7)
- https://www.oracle.com/java/technologies/security-in-java.html

**Comment by le2sky** [comment]
자바 소스코드를 작성하고, 컴파일을 하면 .class 파일이 생성됩니다.
java 명령어로 바이트코드를 실행하면, JVM은 클래스를 로더를 통해 클래스 파일을 읽어들입니다.
클래스 로더는 클래스 파일 정보를 메모리에 올리고, 검증하며 스태틱 변수를 초기화하는 역할을 합니다.
JVM의 목적은 바이트 코드를 기계어로 번역하여, CPU에게 일을 시키는 것입니다.
이 핵심적인 일을 인터프리터가 수행하게 됩니다.

**Comment by GIVEN53** [comment]
.java -> 컴파일 -> .class(바이트코드) -> JVM 로드 ->  클래스 로더 -> 링킹 -> 초기화 -> 실행
